### ISSUE TITLE
Improve error handling in the sync process.

### ISSUE THREAD

**Issue by iamdharmesh** [woocommerce/woocommerce-square#248]
Currently, when the sync process runs, if any error occurs in the Square API request, the sync fails (except in some cases, such as a `429` error). For example, if the upsert catalog API request to Square fails due to an issue with one product in the batch, the entire sync is marked as failed. This issue has been created to improve error handling, allow the sync process to continue, and provide an alert for the product with the error, instead of failing the entire sync.

The Square API currently does not provide detailed information about products with invalid data, so we need to figure out how to identify the problematic product and skip it.

---

## Product Requirements Section

**Improve error handling in the sync process**

> ***Dependency note:** This issue and* [SQUARE-143](https://linear.app/a8c/issue/SQUARE-143/feature-request-square-if-product-has-an-issue-skip-the-product) *address the same root cause at different scopes and should be solved in the same changeset.* [SQUARE-143](https://linear.app/a8c/issue/SQUARE-143/feature-request-square-if-product-has-an-issue-skip-the-product) *covers product-level skipping; this covers the broader API error handling that makes it possible.*

### **Problem & Expected Outcome**

**The Problem:** When a Square API request fails for any reason within a batch, the entire sync is marked as failed. The plugin provides no information about which product caused the failure, and the Square API itself doesn't always identify the problematic item.

**The Fix:** As a store manager, an API error affecting one item in a batch should not fail the entire sync. Errors should be isolated, logged, and the sync should continue.

### **Acceptance Criteria**

* API errors are caught at the item level where possible, not the batch level
* When an entire batch request fails and the problem item cannot be identified, the sync logs the full error and continues to the next batch rather than stopping
* No sync failure results in a silent outcome — every failure produces a log entry with enough detail to diagnose
* Edge/failure state: repeated API errors across multiple batches are logged individually; the sync completes with a "completed with errors" status

### **Out of Scope**

* Identifying the specific bad item within a failed batch when the Square API does not provide that information (best effort only)
* Changes to retry timing or backoff strategy
* [SQUARE-143](https://linear.app/a8c/issue/SQUARE-143/feature-request-square-if-product-has-an-issue-skip-the-product) (product-level skip behavior) — must be solved in the same changeset

---

**Project classification:** Must Have

Review the [technical implementation document](<https://docs.google.com/document/d/1Uo6I77ch3Dh4-1bxY0X4_XELBczbCcswjheNI9pT-vY/edit?tab=t.0>) before starting work.
